src/psp/guest_request.rs

```diff
@@ -0,0 +1,531 @@
+                return false;
+            }
+            // Set encrypt operation
+            let mut ret: i32 = EVP_EncryptInit_ex(
+                self.ctx_req,
+                EVP_aes_256_gcm(),
+                ptr::null_mut(),
+                ptr::null(),
+                ptr::null(),
+            );
+            if ret != 1 {
+                prints!("ERROR: EVP_EncryptInit_ex failed ({})\n", ret);
+                return false;
+            }
+            // Provide key and iv lengths
+            ret = EVP_CIPHER_CTX_set_key_length(self.ctx_req, KEY_SIZE.try_into().unwrap());
+            if ret != 1 {
+                prints!(
+                    "ERROR: Failed to set the key length for encryption ({})\n",
+                    ret
+                );
+                return false;
+            }
+            ret = EVP_CIPHER_CTX_ctrl(
+                self.ctx_req,
+                EVP_CTRL_GCM_SET_IVLEN.try_into().unwrap(),
+                DEFAULT_IV_SIZE.try_into().unwrap(),
+                ptr::null_mut(),
+            );
+            if ret != 1 {
+                prints!("ERROR: EVP_CIPHER_CTX_ctrl failed ({})\n", ret);
+                return false;
+            }
+            // Provide key and iv
+            let svsm_secrets_va: VirtAddr = pgtable_pa_to_va(PhysAddr::new(svsm_secrets_page));
+            let svsm_secrets: *const SnpSecrets = svsm_secrets_va.as_ptr();
+            let key = *(&(*svsm_secrets).vmpck0 as *const _ as *const [u8; KEY_SIZE]);
+            ret = EVP_EncryptInit_ex(
+                self.ctx_req,
+                ptr::null(),
+                ptr::null_mut(),
+                &key as *const _ as *const u8,
+                &iv as *const _ as *const u8,
+            );
+            if ret != 1 {
+                prints!("ERROR: EVP_EncryptInit_ex failed ({})\n", ret);
+                return false;
+            }
+            // Provide AAD
+            let mut len: i32 = 0;
+            ret = EVP_EncryptUpdate(
+                self.ctx_req,
+                ptr::null_mut(),
+                ptr::addr_of_mut!(len),
+                &(*self.req).hdr.algo as *const u8,
+                aad_len.try_into().unwrap(),
+            );
+            if ret != 1 {
+                prints!("ERROR: Failed to provide AAD for encryption ({})\n", ret);
+                return false;
+            }
+            // Provide plaintext
+            ret = EVP_EncryptUpdate(
+                self.ctx_req,                                          // [in] ctx
+                (*self.req).payload.as_mut_ptr(),                      // [out] ciphertext
+                ptr::addr_of_mut!(len),                                // [out] ciphertext len
+                plaintext.as_ptr(),                                    // [in] plaintext
+                core::mem::size_of_val(plaintext).try_into().unwrap(), // [in] plaintext size
+            );
```

Comment:
Is there any security concern around performing the encryption processing directly in the shared memory? Would it be safer to perform everything in an encrypted page and copy the result to the shared page?

Reply:
I agree, I did that. Thanks.